engine/googletest_result.cpp

```diff
@@ -0,0 +1,520 @@
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "engine/googletest_result.hpp"
+
+#include <cstdlib>
+#include <fstream>
+#include <regex>
+#include <utility>
+
+#include "engine/exceptions.hpp"
+#include "model/test_result.hpp"
+#include "utils/fs/path.hpp"
+#include "utils/format/macros.hpp"
+#include "utils/optional.ipp"
+#include "utils/process/status.hpp"
+#include "utils/sanity.hpp"
+#include "utils/text/exceptions.hpp"
+#include "utils/text/operations.ipp"
+
+namespace fs = utils::fs;
+namespace process = utils::process;
+namespace text = utils::text;
+
+using utils::none;
+using utils::optional;
+
+
+namespace {
+
+/// Parses a test result that does not accept a reason.
+///
+/// \param status The result status name.
+/// \param rest The rest of the line after the status name.
+///
+/// \return An object representing the test result.
+///
+/// \throw format_error If the result is invalid (i.e. rest is invalid).
+///
+/// \pre status must be "successful".
+static engine::googletest_result
+parse_without_reason(const std::string& status, const std::string& rest)
+{
+    if (!rest.empty())
+        throw engine::format_error(F("%s cannot have a reason") % status);
+
+    if (status == "skipped")
+        return engine::googletest_result(engine::googletest_result::skipped);
+    else {
+        INV(status == "successful");
+        return engine::googletest_result(engine::googletest_result::successful);
+    }
+}
+
+
+/// Parses a test result that needs a reason.
+///
+/// \param status The result status name.
+/// \param rest The rest of the line after the status name.
+///
+/// \return An object representing the test result.
+///
+/// \throw format_error If the result is invalid (i.e. rest is invalid).
+///
+/// \pre status must be one of "broken", "disabled", "failed", or "skipped".
+static engine::googletest_result
+parse_with_reason(const std::string& status, const std::string& rest)
+{
+    using engine::googletest_result;
+
+    INV(!rest.empty());
+
+    if (status == "broken")
+        return googletest_result(googletest_result::broken, rest);
+    else if (status == "disabled")
+        return googletest_result(googletest_result::disabled, rest);
+    else if (status == "failed")
+        return googletest_result(googletest_result::failed, rest);
+    else if (status == "skipped")
+        return googletest_result(googletest_result::skipped, rest);
+    else
+        PRE_MSG(false, "Unexpected status");
+}
+
+
+/// Formats the termination status of a process to be used with validate_result.
+///
+/// \param status The status to format.
+///
+/// \return A string describing the status.
+static std::string
+format_status(const process::status& status)
```

Comment:
There is a ton of copy/paste in this change from many different sources.

In most cases it seems hard to avoid, but I don't see why this specific function needs duplication. Would be good to reconsider others as well.

Reply:
I'm going to push back on this.

The plain and TAP engines don't handle this particular set of behaviors -- only the ATF engine does. It makes sense to consolidate this in a common area, but that could/should be done as part of a separate commit.